### PR TITLE
Test Menu options for Country selection and Measurement (EXPOSUREAPP-2229)

### PR DESCRIPTION
**Pull request by mertsafter**
## Description
Some implementations for interoperability
* **Country Selection** in Test Menu
* Measurement of **Diagnostic Key Retrieval** and **Risk Calculation** in Test Menu
* Usage of ApplicationConfiguration for country codes **~~(TODO: Need to align property name with backend)~~**

### CODE REVIEW

**Comment by MertSafter** [comment]
Had to adjust the unit tests to respect the country filter. Also there were some linting issues, because of that the build failed. Should be fixed now.

File: Corona-Warn-App/src/main/res/values-bg/strings.xml [hunk 1]
```diff
@@ -1125,5 +1125,6 @@
     <string name="test_api_body_other_keys">"Other key"</string>
     <!-- NOTR -->
     <string name="test_api_calculate_risk_level">"Calculate Risk Level"</string>
+    <string name="test_api_country_filter_status">Country filter applied for: \n%1$s</string>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/values-bg/strings.xml:1128]: Why are we adding english strings to the localized files here?
Wouldn't just adding it to the default be enough as we have no translations for this?
Is this related to the translation setup?

File: Corona-Warn-App/src/main/res/values-de/strings.xml [hunk 2]
```diff
@@ -1140,5 +1140,6 @@
     <string name="test_api_body_other_keys">"Other key"</string>
     <!-- NOTR -->
     <string name="test_api_calculate_risk_level">"Calculate Risk Level"</string>
+    <string name="test_api_country_filter_status">Country filter applied for: \n%1$s</string>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/values-de/strings.xml:1143]: Unlocalized?

File: Corona-Warn-App/src/main/res/values-pl/strings.xml [hunk 3]
```diff
@@ -1125,5 +1125,6 @@
     <string name="test_api_body_other_keys">"Other key"</string>
     <!-- NOTR -->
     <string name="test_api_calculate_risk_level">"Calculate Risk Level"</string>
+    <string name="test_api_country_filter_status">Country filter applied for: \n%1$s</string>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/values-pl/strings.xml:1128]: Unlocalized?

File: Corona-Warn-App/src/main/res/values-ro/strings.xml [hunk 4]
```diff
@@ -1125,5 +1125,6 @@
     <string name="test_api_body_other_keys">"Other key"</string>
     <!-- NOTR -->
     <string name="test_api_calculate_risk_level">"Calculate Risk Level"</string>
+    <string name="test_api_country_filter_status">Country filter applied for: \n%1$s</string>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/values-ro/strings.xml:1128]: Unlocalized?

File: Corona-Warn-App/src/main/res/values-tr/strings.xml [hunk 5]
```diff
@@ -1125,5 +1125,6 @@
     <string name="test_api_body_other_keys">"Other key"</string>
     <!-- NOTR -->
     <string name="test_api_calculate_risk_level">"Calculate Risk Level"</string>
+    <string name="test_api_country_filter_status">Country filter applied for: \n%1$s</string>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/values-tr/strings.xml:1128]: Unlocalized?

File: Corona-Warn-App/src/main/res/values-de/strings.xml (hunk 2 above)
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-de/strings.xml:1143]: Please only add strings to values-de/strings.xml. And add the corresponding comments for the translation system. In this case it would be __.

As that is not enough you also have to add only a placeholder string to values/strings.xml. 
`<string name="test_api_country_filter_status" />`

Please remove the strings for all other languages. To clarify this process I'll send you the link to our internal wiki.

File: Corona-Warn-App/src/main/res/values-bg/strings.xml (hunk 1 above)
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-bg/strings.xml:1128]: please remove

File: Corona-Warn-App/src/main/res/values-de/strings.xml (hunk 2 above)
Comment (harambasicluka) [reply Corona-Warn-App/src/main/res/values-de/strings.xml:1143]: Also the other string is missing `test_api_country_settings_heading`.

File: Corona-Warn-App/src/main/res/values-en/strings.xml [hunk 6]
```diff
@@ -1125,5 +1125,7 @@
     <string name="test_api_body_other_keys">"Other key"</string>
     <!-- NOTR -->
     <string name="test_api_calculate_risk_level">"Calculate Risk Level"</string>
+    <string name="test_api_country_filter_status">Country filter applied for: \n%1$s</string>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-en/strings.xml:1128]: please remove

File: Corona-Warn-App/src/main/res/values-pl/strings.xml (hunk 3 above)
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-pl/strings.xml:1128]: please remove

File: Corona-Warn-App/src/main/res/values-ro/strings.xml (hunk 4 above)
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-ro/strings.xml:1128]: please remove

File: Corona-Warn-App/src/main/res/values-en/strings.xml [hunk 7]
```diff
@@ -1125,5 +1125,7 @@
...
+    <string name="test_api_country_settings_heading">Country Settings</string>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-en/strings.xml:1129]: please remove

File: Corona-Warn-App/src/main/res/values-tr/strings.xml (hunk 5 above)
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-tr/strings.xml:1128]: please remove

File: Corona-Warn-App/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1143,5 +1143,7 @@
     <string name="test_api_calculate_risk_level">"Calculate Risk Level"</string>
     <!-- NOTR -->
     <string name="test_api_switch_background_notifications">"Background Notifications"</string>
+    <string name="test_api_country_filter_status">Country filter applied for: \n%1$s</string>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/strings.xml:1146]: please add comment and empty strings.

File: Corona-Warn-App/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -1143,5 +1143,7 @@
...
+    <string name="test_api_country_settings_heading">Country Settings</string>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/strings.xml:1147]: please add comment and empty strings.

**Review by harambasicluka — CHANGES_REQUESTED** [review]
Small string changes required.

**Review by d4rken — CHANGES_REQUESTED** [review]
Looks mostly good, a few extra changes here and there as we should leave the code always a bit better, sprinkle in some refactoring :heart: .

File: Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/TestForAPIFragment.kt [hunk 10]
```diff
@@ -46,35 +47,12 @@ import de.rki.coronawarnapp.storage.AppDatabase
...
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.label_exposure_summary_summationRiskScore
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.label_googlePlayServices_version
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.label_latest_key_date
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.label_my_keys
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.qr_code_viewpager
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.test_api_switch_last_three_hours_from_server
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.test_api_switch_background_notifications
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.text_my_keys
-import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.text_scanned_key
+import kotlinx.android.synthetic.deviceForTesters.fragment_test_for_a_p_i.*
```
[the remark this answers is not in the document]
Comment (MertSafter) [reply Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/TestForAPIFragment.kt:55]: Refactoring done. Will be in the next commit

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/FileStorageHelper.kt [hunk 11]
```diff
@@ -36,7 +36,7 @@ object FileStorageHelper {
      */
     val keyExportDirectory = File(
         CoronaWarnApplication.getAppContext().cacheDir,
-        FileStorageConstants.KEY_EXPORT_DIRECTORY_NAME
+        "${FileStorageConstants.KEY_EXPORT_DIRECTORY_NAME}/"
```
[the remark this answers is not in the document]
Comment (MertSafter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/FileStorageHelper.kt:39]: I used it to check an issue in the file creation for the key files. Wasn't meant to be commited. Is removed again. Thanks 👍

File: Corona-Warn-App/src/main/res/values-de/strings.xml (hunk 2 above)
Comment (MertSafter) [reply Corona-Warn-App/src/main/res/values-de/strings.xml]: Already discussed with @harambasicluka and fixed

File: Corona-Warn-App/src/main/res/values-bg/strings.xml (hunk 1 above)
Comment (MertSafter) [reply Corona-Warn-App/src/main/res/values-bg/strings.xml]: Maybe I think is it better if we remove this completely from the translation setup

File: Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/TestForAPIFragment.kt [hunk 12]
```diff
@@ -283,6 +305,123 @@ class TestForAPIFragment : Fragment(), InternalExposureNotificationPermissionHel
         updateExposureSummaryDisplay(null)
     }
 
+    /**
+     * Calls the RetrieveDiagnosisKeysTransaction and RiskLevelTransaction and measures them.
+     * Results are displayed using a label
+     * @param callCount defines how often the transactions should be called (each call will be
+     * measured separately)
+     */
+    private suspend fun measureRiskLevelAndKeyRetrieval(callCount: Int) {
```
[the remark this answers is not in the document]
Comment (MertSafter) [reply Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/TestForAPIFragment.kt:312]: Good Idea. Is moved to it's own class. Will be in the next commit.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt [hunk 13]
```diff
@@ -44,6 +44,7 @@ import java.util.UUID
  * Singleton used for accessing key files via combining cached entries from existing files and new requests.
  * made explicitly with [asyncFetchFiles] in mind
  */
+@SuppressWarnings("LongMethod")
```
[the remark this answers is not in the document]
Comment (MertSafter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt:47]: I thought, that the if part will be removed as it is written in the TODO, because of that I didn't invested time to split it.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RetrieveDiagnosisKeysTransaction.kt [hunk 14]
```diff
@@ -175,10 +197,24 @@ object RetrieveDiagnosisKeysTransaction : Transaction() {
         } else {
             Timber.tag(TAG).w("no key files, skipping submission to internal API.")
         }
+
+        if (CWADebug.isDebugBuildOrMode) {
+            var totalFileSize: Long = 0
+            val fileSizeUnit = (1024 * 1024)
```
[the remark this answers is not in the document]
Comment (MertSafter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RetrieveDiagnosisKeysTransaction.kt:203]: Implemented.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt:47]: I think the TODO has been there for weeks, looks like it is here to stay due to being used by the `deviceForTesters` build. :grin: After splitting it, you can remove the TODO.

File: Corona-Warn-App/src/main/res/values-bg/strings.xml (hunk 1 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/res/values-bg/strings.xml]: @harambasicluka Thoughts? Should test related strings be hardcoded to not popup in the translation tooling?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt [hunk 15]
```diff
@@ -151,11 +211,12 @@ object CachedKeyFileHolder {
      * Determines whether a given String has an existing date cache entry under a unique name
      * given from the URL that is based on this String
      *
+     * @param country the country of the date
      * @param cache the given cache entries
      */
-    private fun String.dateEntryCacheMiss(cache: List<KeyCacheEntity>) = !cache
+    private fun String.dateEntryCacheMiss(country: String, cache: List<KeyCacheEntity>) = !cache
```
[the remark this answers is not in the document]
Comment (MertSafter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt:217]: Done.

File: Corona-Warn-App/src/main/res/values-bg/strings.xml (hunk 1 above)
Comment (MertSafter) [reply Corona-Warn-App/src/main/res/values-bg/strings.xml]: I've removed the translation in the test fragment, because as I checked how the other labels are done, I was the only one added it to the translation, now it the same as the other labels are done.

**Review by harambasicluka — APPROVED** [review]

File: Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/RiskLevelAndKeyRetrievalBenchmark.kt [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+            var keyRetrievalError = ""
+            var keyFileCount: Int = -1
+            var keyFileDownloadDuration: Long = -1
+            var keyFilesSize: Long = -1
+
+            try {
+                measureDiagnosticKeyRetrieval(
+                    "#$index",
+                    countries
+                ) { duration, keyCount, totalFileSize ->
```
Comment (d4rken) [inline Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/RiskLevelAndKeyRetrievalBenchmark.kt:48]: Is the callback required, can't this just return the 3 values?

File: Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/RiskLevelAndKeyRetrievalBenchmark.kt [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+                }
+            } catch (e: TransactionException) {
+                keyRetrievalError = e.message.toString()
+            }
+
+            var calculationDuration: Long = -1
+            var calculationError = ""
+
+            try {
+                measureKeyCalculation("#$index") {
```
Comment (d4rken) [inline Corona-Warn-App/src/deviceForTesters/java/de.rki.coronawarnapp/RiskLevelAndKeyRetrievalBenchmark.kt:61]: Same as above, is the callback really necessary here?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RetrieveDiagnosisKeysTransaction.kt [hunk 18]
```diff
@@ -175,10 +197,24 @@ object RetrieveDiagnosisKeysTransaction : Transaction() {
         } else {
             Timber.tag(TAG).w("no key files, skipping submission to internal API.")
         }
+
+        if (CWADebug.isDebugBuildOrMode) {
+            var totalFileSize: Long = 0
+            keyFiles.also {
+                it.forEach { file ->
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RetrieveDiagnosisKeysTransaction.kt:204]: Maybe `keyFiles.forEach` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt [hunk 19]
```diff
@@ -67,145 +67,217 @@ object CachedKeyFileHolder {
...
+                        .map { url ->
+                            async {
+                                return@async WebRequestBuilder.getInstance()
+                                    .asyncGetKeyFilesFromServer(url)
+                            }
+                        }
+                }
+                .awaitAll()
+        } else {
+            throw IllegalStateException(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CachedKeyFileHolder.kt:181]: Let's turn this into an early abort:

```kotlin
if (packagesWithCurrentDate.isNotEmpty()) throw Illegal....
```

**Review by d4rken — CHANGES_REQUESTED** [review]

**Review by d4rken — APPROVED** [review]
